ocs_ci/utility/csr.py

```diff
@@ -2,29 +2,42 @@
 
 from ocs_ci.ocs import constants, exceptions, ocp
 from ocs_ci.utility.retry import retry
-from ocs_ci.utility.utils import run_cmd
+from ocs_ci.utility.utils import run_cmd, TimeoutSampler
 
 logger = logging.getLogger(__name__)
 
 
-@retry(exceptions.PendingCSRException, tries=4, delay=10, backoff=1)
-def approve_pending_csr():
+@retry(
+    (exceptions.PendingCSRException, exceptions.TimeoutExpiredError),
+    tries=4,
+    delay=10,
+    backoff=1
+)
+def approve_pending_csr(expected=None):
     """
     After node addition CSR could be in pending state, we have to approve it.
 
+    Args:
+        expected (int): Expected number of CSRs. By default, it will approve
+            all the pending CSRs if exists.
+
     Raises:
         exceptions.PendingCSRException
+        exceptions.TimeoutExpiredError
 
     """
-    base_cmd = "oc adm certificate approve"
-    pending_csrs = get_pending_csr()
-    if pending_csrs:
-        logger.info(f"Pending CSRs: {pending_csrs}")
-        csrs = ' '.join([str(csr) for csr in pending_csrs])
-        cmd = f"{base_cmd} {csrs}"
-        logger.info("Approving pending CSRs")
-        run_cmd(cmd)
-
+    for pending_csrs in TimeoutSampler(300, 10, get_pending_csr):
+        if not expected:
+            if pending_csrs:
+                approve_csrs(pending_csrs)
+            break
+        if len(pending_csrs) >= expected:
```

Comment:
I don't think it's enough as I could see about 47 CSRs.
```
Nodes CSRS: ['CSR: csr-6hx8z , Node: {node_info[1]}', 'CSR: csr-blz2n , Node: {node_info[1]}', 'CSR: csr-p46bz , Node: {node_info[1]}', 'CSR: csr-tzhn5 , Node: {node_info[1]}', 'CSR: csr-wm4g5 , Node: {node_info[1]}']
Out[29]: 5

In [30]: nodes_csrs = []
    ...: for item in data['items']:
    ...:
    ...:     print(item['metadata']['name'])
    ...:     status = item.get('status')
    ...:     print(f"status: {status['conditions'][0]['type']}")
    ...:     node_info = item['spec'].get("username").split('node:')
    ...:     if len(node_info) == 2:
    ...:         print(f"Node: {node_info[1]}")
    ...:         nodes_csrs.append(f"CSR: {item['metadata']['name']} , Node: {{node_info[1]}}")
    ...: print(f"Nodes CSRS: {nodes_csrs}")
    ...: len(nodes_csrs)
```
I think we should have other function which will check that for all the nodes we have from `oc get nodes` that each of them has approved CSR . WDYT?

Reply:
Just adding pastebins from our hangout chat also here:
Ipython output: http://pastebin.test.redhat.com/859321
Data : http://pastebin.test.redhat.com/859322

log:
http://magna002.ceph.redhat.com/ocsci-jenkins/openshift-clusters/jnk-vu1rs33-t4b/jnk-vu1rs33-t4b_20200427T114109/logs/ocs-ci-logs-1587988369/by_outcome/failed/tests/ecosystem/deployment/test_deployment.py/test_deployment/logs
Jenkins job: 
https://ocs4-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/qe-deploy-ocs-cluster/7086/console

Reply:
@petr-balogh ,
While testing adding RHCOS nodes I am facing same problem. Another issue here is , while a node has pending csr, it doesn't get listed in ```oc get nodes`` and I am not sure even if we check for every node in the output of ```oc get nodes``` can be foolproof.

I mean there is a small window where 
1. There won't be any pending csr in ```oc get csr```
2. All the nodes from ```oc get nodes``` would be in ready state.
BUT still new node hasn't appeared . Not sure how precisely we can catch this.

Reply:
Another solution I can think of is 
1. Get the IP of the nodes added and keep on doing csr approval till it appears in ```oc get nodes``` output

Reply:
Another solution I can think of is , I somewhere read, Leaving the boostrap instances as is after deployment.
If bootstrap instance is alive it can take care of approving csrs automatically though I am not 100% sure here.

Reply:
I am not sure we can leave bootstrap, I remember we need to remove bootstrap as part of deployment. Added few more functions to wait till all nodes generated CSRs